katdal/applycal.py

```diff
@@ -86,17 +86,6 @@ def complex_interp(x, xi, yi, left=None, right=None):
     return y.astype(yi.dtype)
 
 
-def has_cal_product(cache, attrs, product):
```

Comment:
I have a question on this...

I removed this because I could not check `Calibration/Products` virtual sensors. I still have the opportunity to scan through the keys of the underlying stream(s) and find all products up front though, which could be useful for debugging / diagnostics. This would be simpler if `telstate.keys()` were view-aware instead of listing the global keys (rather like `telstate._ipython_key_completions_` 😄). What was the reasoning behind this again?

Reply:
I think we just decided that there was no obvious right answer for what "view-aware" would mean. In particular, in a telstate with views, some keys could be inaccessible because they're shadowed, while others may be reachable via multiple names.

But I think `key in telstate` should be view-aware, which should be all you need, right?

Reply:
I was thinking of iterating over the keys and finding all the ones that start with `product_`, but I would have done this on an exclusive "cal" view so there is no issue of shadowing or duplicates.

Reply:
With an exclusive view I guess it would be well-defined.

I would hesitate to change the behaviour of `keys` at this point (it'll be very difficult to find all the places it gets called), but maybe it would make sense to introduce another function that has behaviour similar to the IPython completer that could be used for this.